tests/components/zha/test_discover.py

```diff
@@ -30,45 +66,298 @@
         node_descriptor=device["node_descriptor"],
     )
 
-    _dispatch = mock.MagicMock(wraps=disc.async_dispatch_discovery_info)
-    monkeypatch.setattr(core_zha_gw, "async_dispatch_discovery_info", _dispatch)
-    entity_registry = await homeassistant.helpers.entity_registry.async_get_registry(
-        hass
+    orig_new_entity = zha_channels.ChannelPool.async_new_entity
+    _dispatch = mock.MagicMock(wraps=orig_new_entity)
+    try:
+        zha_channels.ChannelPool.async_new_entity = lambda *a, **kw: _dispatch(*a, **kw)
+        zha_dev = await zha_device_joined_restored(zigpy_device)
+        await hass.async_block_till_done()
+    finally:
+        zha_channels.ChannelPool.async_new_entity = orig_new_entity
+
+    entity_ids = hass.states.async_entity_ids()
+    await hass.async_block_till_done()
+    zha_entity_ids = {
+        ent for ent in entity_ids if ent.split(".")[0] in zha_const.COMPONENTS
+    }
+
+    event_channels = {
+        ch.id for pool in zha_dev.channels.pools for ch in pool.relay_channels.values()
+    }
+
+    entity_map = device["entity_map"]
+    assert zha_entity_ids == set(
+        [
+            e["entity_id"]
+            for e in entity_map.values()
+            if not e.get("default_match", False)
+        ]
     )
+    assert event_channels == set(device["event_channels"])
+
+    for call in _dispatch.call_args_list:
+        _, component, entity_cls, unique_id, channels = call[0]
+        key = (component, unique_id)
+        entity_id = entity_registry.async_get_entity_id(component, "zha", unique_id)
+
+        assert key in entity_map
+        assert entity_id is not None
+        no_tail_id = NO_TAIL_ID.sub("", entity_map[key]["entity_id"])
+        assert entity_id.startswith(no_tail_id)
+        assert set([ch.name for ch in channels]) == set(entity_map[key]["channels"])
+        assert entity_cls.__name__ == entity_map[key]["entity_class"]
+
+
+@mock.patch(
+    "homeassistant.components.zha.core.discovery.ProbeEndpoint.discover_by_device_type"
+)
+@mock.patch(
+    "homeassistant.components.zha.core.discovery.ProbeEndpoint.discover_by_cluster_id"
+)
+def test_discover_entities(m1, m2):
+    """Test discover endpoint class method."""
+    ep_channels = mock.MagicMock()
+    disc.PROBE.discover_entities(ep_channels)
+    assert m1.call_count == 1
+    assert m1.call_args[0][0] is ep_channels
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5051600)